### DISCUSSION TITLE
Added: LDM Notes for January 12th, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5646] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-01-12.md

## Agenda

1. Open questions for `field`
    1. Initializers for semi-auto properties
    2. Definite assignment for struct types
2. Generic Math Operator Enhancements

**Comment by CyrusNajmabadi** [comment]
Note: additional information on:

> 3. Allow all properties with backing fields to have initializers. These would assign to the backing field, and there would be a meaningful difference between the initializer and assigning in the constructor.
> 
> After some discussion, we feel that 3 is a step too far. 

It's worth pointing out that there *can* be a meaningful difference today with properties and assigning in the initializer vs assigning in the constructor.  For example, a virtual auto-property will assign directly into the field with an initializer, but will dispatch through the setter if in the constructor.  This can be observable depending on what subclasses do.

That said, this is def a corner case, and not likely to happen much in practice.  If we were to allow semi-auto-props to have this distinct behavior, this difference would become a lot more common.  What was currently an unlikely, and thus ok to be unpalatable, case would now become very common and thus potentially more confusing and undesirable.

> It could potentially be enabled later, if we can find a set of rules that makes sense and unifies the area, but for the moment we don't have those rules.

I agree with this.  It would not conflict with teh more conservative rules we're currently going with and we can see how unpleasant not having this is for the community.

**Comment by jnm2** [comment]
> 3. Allow all properties with backing fields to have initializers. These would assign to the backing field, and there would be a meaningful difference between the initializer and assigning in the constructor.
>
> After some discussion, we feel that 3 is a step too far. It could potentially be enabled later, if we can find a set of rules that makes sense and unifies the area, but for the moment we don't have those rules.

This immediately penalizes all INPC properties that have non-type-default initial property values. Refactoring to needing such an initial value pushes them over a cliff into not being able to use `field`. It's a frequent scenario to have variations on:

```cs
private bool isActive = true;
public bool IsActive { get => isActive; set => Set(ref isActive, value); }
```

What I've been hoping for, and this is completely consistent with today's observable property initializer behavior on virtual auto properties:
```cs
public bool IsActive { get; set => Set(ref field, value); } = true;
```

The notes go on to say that the property-scoped fields proposal could address this pain. But even if it was available from day 1, this end result is disappointing:
```cs
public bool IsActive { bool field = true; get => field; set => Set(ref field, value); }
```

**Reply by 333fred** [reply]
> This immediately penalizes all INPC properties that have non-type-default initial property values.

My belief is that this is not true, though investigating the struct problem may well render the problem moot. The constructor will assign the non-default value, and the setter will run. Nothing is listening for INPC, so no updates will be fired.

**Reply by jnm2** [reply]
@333fred That's not necessarily good enough. Consider a case where setting one property (intended to represent data binding from a UI edit) changes another property or even kicks off work:

```cs
public int Foo
{
    get;
    set
    {
        if (!Set(ref field, value)) return;
        // Stuff, e.g.: IsModified = true.
        // Using the constructor to initialize is now messy.
    }
}
```

**Reply by jnm2** [reply]
Not to mention, placing the default value in the constructor takes it far away from the property declaration in most cases.

**Reply by sab39** [reply]
What if there were a syntax like this:

```csharp
public bool IsActive { get; set => Set(ref field, value); field = true; }
```

**Reply by CyrusNajmabadi** [reply]
That is effectively the 'locals in property' proposal which we like, but are scoping out of this initial work/proposal to keep costs down.

**Reply by sab39** [reply]
Yes, except it's referencing the existing generated special field rather than declaring a separate local, so that `get` can still be bodyless.

**Reply by sab39** [reply]
To elaborate a little: my suggestion is to allow specifically the context-keyword `field` to be used in this way, so the only things that could appear in a property block would be `get`, `set`, `init` and `field =`. As best I can tell that would be consistent with allowing generalized locals in properties in the future, but wouldn't significantly increase the scope of _this_ feature because it would merely provide a syntax for initializers of the synthesized backing field, which already exists for regular auto-properties.

**Comment by sab39** [comment]
Does the current proposal allow something like the following for lazy initialization? I suspect that as I've written it here it would run afoul of definite assignment and nullability rules, but it'd definitely be handy to have a concise way to do something like this:

```csharp
public Foo Foo => field ??= initializeFoo();
```

(edited because I'd forgotten that `??=` was a thing)

**Reply by 333fred** [reply]
We think we should be able to do a bit of inference and figure out that `field` should be nullable for these lazy cases, and require you to do that initialization if it hasn't already been done.
